### PR TITLE
DR-3251: Refactor API handler, fix timeout error

### PR DESCRIPTION
**Pull request by 7emansell**
## Ticket:

Resolves JIRA ticket [DR-3251](https://newyorkpubliclibrary.atlassian.net/browse/DR-3251)

## This PR does the following:

- Refactors API handlers into one catch-all function, `apiResponse`
    - And moves testing around accordingly
- Adds timeout error handling
** Hard to test with certainty (see [here](https://github.com/NYPL/digital-collections/blob/d79bf1758bfa8d76e44b0612ab9bb365a7546a79/app/src/utils/api.test.tsx#L162), I'm just not very confident in that). I'll ask Jeff to recreate the Repo API crash once this is on QA and that will confirm whether this actually works
- Unrelated but also adds `--watch` flag as `npm run test:watch`

## Open Questions

- Is 7 seconds the right amount of time to wait?
- Answer: Repo API's slowest call should be about 6 seconds, so yes

### Checklist:

- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.
- [x] I have updated the CHANGELOG.md.

[DR-3251]: https://newyorkpubliclibrary.atlassian.net/browse/DR-3251?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by EdwinGuzman** [comment]
> Is 7 seconds the right amount of time to wait?

This is a good question for the backend team working on the Repo API. If they have a timeout for a request/response of x seconds, the frontend should have that value as a minimum. Otherwise, this might assume that the response is timing out while it's actually still working.

File: app/src/utils/api.test.tsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    await expect(apiResponse(mockApiUrl)).rejects.toEqual(
+      new Error("apiResponse: 500 Internal Server Error")
+    );
+  });
+
+  it("throws a timeout error if the request takes too long", async () => {
+    jest.useFakeTimers();
+
+    (fetch as jest.Mock).mockImplementation(() => new Promise(() => {}));
```
Comment (EdwinGuzman) [inline app/src/utils/api.test.tsx:165]: I was trying to run this test in isolation (using `it.only`) but it fails with the following:

Screenshot 2024-11-05 at 8 19 43 PM

Follow the pattern from the previous test to mock fetch. Ideally, tests should not rely on global mocks or vars.

**Review by EdwinGuzman — COMMENTED** [review]
Solid refactor and wrapper for catching the API timeout calls. Left some minor comments.

We've been discussing Fowler's Refactoring book and what comes to mind is that existing unit tests still work with this update. I know you reworked and moved some tests around, but do you think the existing tests were "good enough" to prove the point above?

**Comment by 7emansell** [comment]
> This is a good question for the backend team working on the Repo API. If they have a timeout for a request/response of x seconds, the frontend should have that value as a minimum. Otherwise, this might assume that the response is timing out while it's actually still working.

I'll double check with Kris Kelly tomorrow but no one else was aware of a specific timeout sooner than 60 seconds (which seems like not what we want even if Repo API is okay with it?)

File: app/src/utils/api.ts [hunk 2]
```diff
@@ -140,111 +143,93 @@ export const getItemByIdentifier = async (
...
+    apiUrl += queryString;
   }
-};
 
-/**
- * Makes a POST request to Repo API.
- */
-export const apiPOSTCall = async (apiUrl: string, postData: any) => {
-  const apiKey = process.env.AUTH_TOKEN;
+  const timeout = 7000;
```
[the remark this answers is not in the document]
Comment (7emansell) [reply app/src/utils/api.ts:200]: Totally just random personal preference

**Comment by 7emansell** [comment]
> Solid refactor and wrapper for catching the API timeout calls. Left some minor comments.
> 
> We've been discussing Fowler's Refactoring book and what comes to mind is that existing unit tests still work with this update. I know you reworked and moved some tests around, but do you think the existing tests were "good enough" to prove the point above?

No, the existing tests were for sure not convincing coverage (I'm not sure what you mean by the "point above" but I think you mean accurate coverage).  Although to be fair I was basically writing new tests for `api.ts`– now errors are being thrown, and I wanted to test how `apiReponse` constructs requests, and there was no coverage for either in the existing tests. All of the other functions fetching data (`getHomepageData`, `getNumDigitizedItems`) should have unit tests though. Do we want to consider that part of the scope of this ticket?

**Comment by EdwinGuzman** [comment]
> I'm not sure what you mean by the "point above" but I think you mean accurate coverage

The point is that tests passed before, you made a big refactor, and tests still pass. But you're right that tests were not fully covering everything before and you added new tests.

> Do we want to consider that part of the scope of this ticket? @EdwinGuzman 

No. I think this is good to have QA test for timeouts but do create a ticket for writing more unit tests. (But perhaps get some input from others regarding the timeout value).

**Review by EdwinGuzman — APPROVED** [review]

**Comment by 7emansell** [comment]
Ticket for full test coverage: https://newyorkpubliclibrary.atlassian.net/browse/DR-3271

**Review by avertrees — APPROVED** [review]

**Pull request merged**